contracts/PermissionLessReserve.sol

```diff
@@ -0,0 +1,596 @@
+
+        //notice. if decimal API not supported this should revert
+        setDecimals(reserveToken);
+        require(getDecimals(reserveToken) <= MAX_DECIMALS);
+    }
+
+    function getConversionRate(ERC20 src, ERC20 dest, uint totalSrcAmount, uint blockNumber) public view returns(uint) {
+
+        require((src == ETH_TOKEN_ADDRESS) || (dest == ETH_TOKEN_ADDRESS));
+        require((src == reserveToken) || (dest == reserveToken));
+        blockNumber; // in this reserve no order expiry == no use for blockNumber. here to avoid compiler warning.
+
+        uint32 orderId;
+
+        if (src == ETH_TOKEN_ADDRESS) {
+            if (isNextOrderTail(BUY_HEAD_ID)) return 0;
+            orderId = BUY_HEAD_ID;
+        } else {
+            if (isNextOrderTail(SELL_HEAD_ID)) return 0;
+            orderId = SELL_HEAD_ID;
+        }
+
+        uint128 remainingSrcAmount = uint128(totalSrcAmount);
+        uint128 totalDstAmount = 0;
+
+        while (!isNextOrderTail(orderId)) {
+
+            orderId = getNextOrderId(orderId);
+
+            Order memory order = orders[orderId];
+
+            if (order.srcAmount < remainingSrcAmount) {
+                totalDstAmount += order.dstAmount;
+                remainingSrcAmount -= order.srcAmount;
+            } else {
+                totalDstAmount += order.dstAmount * remainingSrcAmount / order.srcAmount;
+                remainingSrcAmount = 0;
+                break;
+            }
+        }
+
+        if ((remainingSrcAmount != 0) || (totalDstAmount == 0)) return 0; //not enough tokens to exchange.
+
+        //check overflow
+        if (uint(totalDstAmount) * PRECISION < uint(totalDstAmount)) return 0;
+
+        return calcRateFromQty(totalSrcAmount, totalDstAmount, getDecimals(src), getDecimals(dest));
+    }
+
+    function trade(
+        ERC20 srcToken,
+        uint srcAmount,
+        ERC20 destToken,
+        address destAddress,
+        uint conversionRate,
+        bool validate
+    )
+        public
+        payable
+        returns(bool)
+    {
+        require((srcToken == ETH_TOKEN_ADDRESS) || (destToken == ETH_TOKEN_ADDRESS));
+        require((srcToken == reserveToken) || (destToken == reserveToken));
+
+        if (validate) {
+            require(conversionRate > 0);
+        }
+
+        uint32 orderId;
+
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            require(msg.value == srcAmount);
+            orderId = BUY_HEAD_ID;
+        } else {
+            require(srcToken.transferFrom(msg.sender, this, srcAmount));
+            require(msg.value == 0);
+            orderId = SELL_HEAD_ID;
+        }
+
+        uint128 remainingSrcAmount = uint128(srcAmount);
+        uint128 totalDstAmount = 0;
+
+        while (!isNextOrderTail(orderId)) {
+
+            // start with next since head order is dummy order
+            orderId = getNextOrderId(orderId);
+
+            Order memory order = orders[orderId];
+
+            if (order.srcAmount <= remainingSrcAmount) {
```

Comment:
probably `<` and not `<=`. this way you can avoid `if (remainingSrcAmount == 0) break;`